packages/flutter/lib/src/material/colors.dart

```diff
@@ -88,6 +90,95 @@ class MaterialAccentColor extends ColorSwatch<int> {
   Color get shade700 => this[700];
 }
 
+/// Signature for the function that returns a (potentially different) color
+/// based on a given set of states.
+typedef ColorResolver = Color Function(Set<MaterialState> states);
+
+/// Defines a [Color] that can react to changes in the state of a Material
+/// component, based on a given set of [MaterialState]s.
+///
+/// This is useful for preserving the accessibility of text in different states
+/// of a component. For example, in a [FlatButton] with blue text, the text will
+/// become more difficult to read when the button is hovered, focused, or pressed,
+/// because the contrast ratio between the button and the text will decrease. To
+/// solve this, you can use [MaterialStateColor] to make the text darker when the
+/// [FlatButton] is hovered, focused, or pressed.
+///
+/// To use a [MaterialStateColor], you can either:
+///   1. Create a subclass of [MaterialStateColor] and implement the abstract `resolve` method.
+///   2. Use [MaterialStateColor.resolveWith] and pass in a callback that
+///      will be used to resolve the color in the given states.
+///
+/// This should only be used as parameters when they are documented to take
+/// [MaterialStatefulColor], otherwise only the default state will be used.
+///
+/// Example:
+/// ```dart
+/// Color getTextColor(Set<MaterialSet> states) {
+///   final interactiveStates = <MaterialState>{
+///     MaterialState.pressed,
+///     MaterialState.hovered,
+///     MaterialState.focused,
+///   };
+///   if(states.any(interactiveStates.contains) {
+///     return Colors.blue[900];
+///   }
+///   return Colors.blue[600];
+/// }
+///
+/// FlatButton(
+///   ...
+///   textColor: MaterialStateColor.resolveWith(getTextColor),
+/// ),
+/// ```
+abstract class MaterialStateColor extends Color {
+  /// Creates a [MaterialStateColor], a color that can return different values
+  /// depending on a given [MaterialState].
+  const MaterialStateColor(int defaultValue) : super(defaultValue);
```

Comment:
This constructor is a bit confusing, since it's basically just a regular `Color`. I think the documentation is misleading here. Do you actually need this constructor?

Reply:
Clarified documentation.